Marlin/configuration_store.cpp

```diff
@@ -1729,59 +1848,96 @@ void MarlinSettings::reset() {
     #endif
   );
 
+  #define TMC_INIT_CURRENT(Q) stepper##Q.setCurrent(Q##_CURRENT, R_SENSE, HOLD_MULTIPLIER);
```

Comment:
Reminds me that we could remove `setCurrent` from `tmc_init` functions in `stepper_indirection.cpp` as we already set current (and some other parameters) here already. The newer versions of the libraries support a `push` method that re-sends all the shadow registers to the driver in case the driver was not powered by 12V when initially booting the machine, like when `PS_DEFAULT_OFF` is enabled. But let's not do that yet as it will require testing.

Reply:
As another standard behavior, Marlin initialized all settings-based values in the EEPROM load and reset code, so yes, if they are also being initialized in `stepper_indirection` then we can eliminate the redundancy.

Reply:
Yes, but the `push` command isn't yet even copied into the TMC2208Stepper library.

Reply:
Looks like we will need to add our own "shadow registers" because if you power down with `M81` and then re-power with `M80` the settings will revert to those in the configuration, and won't use the ones saved in the EEPROM.

Reply:
You don't need to start keeping track of the shadow registers when the library is already doing that. No sense duplicating what we already have.

Reply:
I'll check and see if they're exposed as public data members.

Reply:
Darn, they're private. And there's no "shadow" for things like current, anyway.

Reply:
[TMC2130Stepper::push()](https://github.com/teemuatlut/TMC2130Stepper/blob/master/src/source/TMC2130Stepper.cpp#L150)
Write all the shadow registers stored by the library to the driver.

`getCurrent()` returns `val_ma` which is a private variable that holds the user selected current value. 
`rms_current()` returns the actual coil current that's configured in the driver.
Due to the quantization of the driver register CS value it will usually differ from the user's input.

Reply:
Ah, somehow I missed those in scanning the library code.

Reply:
So, let's see… It seems like we will still need to restore the last-set current (and other settings) in response to `M80`. In other words something like `st.setCurrent(st.getCurrent())`. Summing it up…
- On boot or reconnect the `settings.reset` or `settings.load` method restores the steppers to their default or saved configured state.
- We don't need to init all over again in `stepper.setup` due to the first point.
- For `M80` we should try to restore the steppers to the last state they were in, possibly with calls like `st.setCurrent(st.getCurrent())`.

Reply:
I see there is a misspelling in the libraries: The word `hysterisis` should be `hysteresis`.

Reply:
> The word hysterisis should be hysteresis.

I'll fix that but it'll break Marlin for a bit. See `stepper_indirection.cpp`.

`settings.reset` or `settings.load` don't handle all the configuration parameters so we still need to call `init`. Although there may very well still be some that are duplicates.

`M80` can call `st.push()` to sync to the drivers anything that was attempted while there was no 12V power. I'll add this to TMC2208Stepper v0.0.4 as well.

Reply:
> I'll add this to TMC2208Stepper v0.0.4 as well.

Sounds good. I hope we can get that before the 1.1.9 release. In the meantime for this PR I need to come up with some kind of alternative solution.

Note that the `settings.load`/`settings.reset` method is called before the `stepper.init` function. It looks like we'll need to call `stepper.init` first. But I hope there's not some "Catch-22" here, where we can't call `stepper.init` before or during EEPROM loading because it depends on having values already loaded from EEPROM.

Reply:
If `stepper.init` is called before `settings.load` then we no longer have to call `tmc_init_cs_pins` in `setup`.

Reply:
TMC2208Stepper v0.0.4 released.
Adds the `push` method and fixes the same typo that was in TMC2130Stepper.

Reply:
I've updated Marlin to use the latest libraries. I hoped to check the version number in the `SanityCheck.h` but it's not visible to the preprocessor. For the next version that would be a useful addition.